Mathlib/NumberTheory/Zagier.lean

```diff
@@ -0,0 +1,288 @@
+`x * x + 4 * y * z = 4 * k + 1`, as a `Set`. -/
+def zagierSet := {t : ℕ × ℕ × ℕ | t.1 * t.1 + 4 * t.2.1 * t.2.2 = 4 * k + 1}
+
+lemma zagierSet_lower_bound {x y z : ℕ} (h : (x, y, z) ∈ zagierSet k) : 0 < x ∧ 0 < y ∧ 0 < z := by
+  simp_rw [zagierSet, Set.mem_setOf_eq] at h
+  refine' ⟨_, _, _⟩ <;> (by_contra q; push_neg at q; rw [le_zero_iff] at q; subst q; simp at h)
+  · apply_fun (· % 4) at h
+    simp [mul_assoc, Nat.add_mod] at h
+  all_goals cases' (Nat.dvd_prime hk.out).1 (dvd_of_mul_left_eq _ h) with e e <;>
+    (subst e; simp at h; subst h; simp at hk; exact hk.out)
+
+lemma zagierSet_upper_bound {x y z : ℕ} (h : (x, y, z) ∈ zagierSet k) :
+    x ≤ k + 1 ∧ y ≤ k ∧ z ≤ k := by
+  obtain ⟨hx, hy, hz⟩ := zagierSet_lower_bound k h
+  simp_rw [zagierSet, Set.mem_setOf_eq] at h
+  refine' ⟨_, _, _⟩ <;> nlinarith
+
+/-- The set of all triples of natural numbers `(x, y, z) ∈ (0, k + 1] × (0, k] × (0, k]` satisfying
+`x * x + 4 * y * z = 4 * k + 1`, as a `Finset`. This is shown to be equivalent to `zagierSet`. -/
+def zagierFinset :=
+  ((Ioc 0 (k + 1)).product ((Ioc 0 k).product (Ioc 0 k))).filter
+    (fun ⟨x, y, z⟩ => x * x + 4 * y * z = 4 * k + 1)
+
+theorem zagierSet_eq : zagierSet k = zagierFinset k := by
+  ext ⟨x, y, z⟩
+  refine' ⟨fun h => _, fun h => _⟩
+  · unfold zagierSet at h
+    unfold zagierFinset; simp_all
+    have lb := zagierSet_lower_bound k h
+    have ub := zagierSet_upper_bound k h
+    apply mem_product.2; simp
+    constructor; · exact ⟨lb.1, ub.1⟩
+    apply mem_product.2; simp
+    exact ⟨⟨lb.2.1, ub.2.1⟩, ⟨lb.2.2, ub.2.2⟩⟩
+  · unfold zagierFinset at h
+    unfold zagierSet; simp_all
+
+/-- The set of all triples of natural numbers `(x, y, z)` satisfying
+`x * x + 4 * y * z = 4 * k + 1`, as a `Subtype`. -/
+def zagierSubtype := {t // t ∈ zagierSet k}
+
+instance : Fintype (zagierSubtype k) := by
+  unfold zagierSubtype
+  rw [zagierSet_eq]
+  infer_instance
+
+end Defs
+
+section Key
+
+variable {α : Type*} [Fintype α] [DecidableEq α] {p : ℕ} [hp : Fact p.Prime]
+  (f : Function.End α) (hf : f^[p] = id)
+
+open Function Submonoid
+
+/-- A shim theorem. -/
+theorem pow_eq_iterate (k : ℕ) : f ^ k = f^[k] := by
+  induction k with
+    | zero => rfl
+    | succ n ih =>
+      simp only [iterate_succ', ← ih, pow_succ]
+      rfl
+
+/-- The powers of a periodic endomorphism form a group with composition as the operation. -/
+def groupPowers : Group (powers f) where
+  inv := fun ⟨g, hg⟩ => ⟨g^[p - 1], by
+    rw [mem_powers_iff] at hg ⊢
+    obtain ⟨k, hk⟩ := hg
+    use k * (p - 1)
+    rw [← pow_eq_iterate, ← hk, pow_mul]⟩
+  mul_left_inv := fun ⟨g, hg⟩ => by
+    simp only [ge_iff_le, mk_mul_mk]
+    congr
+    rw [← pow_eq_iterate, ← pow_succ',
+      Nat.sub_add_cancel (one_le_two.trans (Nat.Prime.two_le hp.out))]
+    rw [mem_powers_iff] at hg
+    obtain ⟨k, hk⟩ := hg
+    rw [← hk, ← pow_mul, mul_comm, pow_mul, pow_eq_iterate f, hf, pow_eq_iterate, iterate_id]
+    rfl
```

Comment:
I feel like we ought to have this somewhere as the result on arbitrary monoids, that `powers x` is a group if `orderOf x = p` is not zero.

Reply:
I'm not entirely sure about how to effect this. I've added `instGroupOfPowers` in `GroupTheory.Submonoid.Membership`, but it has to be low-priority otherwise `fintypeOfKerOfCodom` in `GroupTheory.QuotientGroup` fails. And what about `isPGroup_of_powers` in the new file?

Reply:
Now that the group construction is no longer needed here it's been moved to #6648.